rgw: Rework of s3 LDAP Authentication code.

pritha-srivastava opened the pull request:
The LDAP authentication code has been reworked based
on the new authentication infrastructure.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava commented:
Things pending:
1. Find appropriate files for class definitions.
2. The old code for ldap handling has not been removed.
3. The ACL bit is a little unclear to me.

mattbenjamin commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -4342,3 +4345,88 @@ RGWOp* RGWHandler_REST_Service_S3Website::get_obj_op(bool get_data)
+{
+  //This is based on the assumption that the default acl strategy in
+  // get_perms_from_aclspec, will take care. Extra acl spec is not required.
+  return nullptr;
```
I think that -should- be true

mattbenjamin commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -4342,3 +4345,88 @@ RGWOp* RGWHandler_REST_Service_S3Website::get_obj_op(bool get_data)
+    }
+  }
+  return token;
+}
```
github is warning--need newline at end of file

mattbenjamin commented:
At a high level, this looks very good.  Class definitions can go wherever, we need to support building with and without LDAP support.  The code not removed is the actual auth hooks?  Does this code actually run, then?  I think the current ACL handling is correct, but :)

pritha-srivastava commented:
@mattbenjamin : This code runs. I replaced the logic in authorize_v2 with something similar to what is there for the swift authorize method and tested it, though I removed the test code before creating the PR.

pritha-srivastava commented:
@mattbenjamin : The code not removed is the piece which is there in authorize_v2. That bit can be removed when we have all the S3 AuthEngines in place. And the hooks like RGW_Auth_S3::init and init_impl and the respective variables.

pritha-srivastava commented:
@mattbenjamin : The code to build with and without ldap support is in rgw_ldap.h ( #if defined(HAVE_OPENLDAP)) . Is this what you were looking for or you meant something else?

pritha-srivastava commented:
Added code for specifying 'type' with the user. Also replaced some parts of the old code with the new AuthEngine and AuthApplier pieces.

mattbenjamin commented on src/rgw/rgw_auth.cc:
```diff
@@ -149,6 +149,11 @@ void RGWRemoteAuthApplier::create_account(const rgw_user& acct_user,
 {
   rgw_user new_acct_user = acct_user;
 
+  if (!info.acct_type.empty()) {
```
cool

mattbenjamin commented on src/rgw/rgw_auth.h:
```diff
@@ -117,16 +117,19 @@ class RGWRemoteAuthApplier : public RGWAuthApplier {
     const std::string acct_name;
     const uint32_t perm_mask;
     const bool is_admin;
+    const std::string acct_type;
```
is std::string the best type for this?  enum class, maybe?

mattbenjamin commented on src/rgw/rgw_common.h:
```diff
@@ -677,6 +680,7 @@ struct RGWUserInfo
     }
     if (struct_v >= 18) {
       ::decode(admin, bl);
+      ::decode(type, bl);
```
for enum class maybe wrap it in a class then use the macros?

mattbenjamin commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -3936,52 +3939,48 @@ int RGW_Auth_S3::authorize_v2(RGWRados *store, struct req_state *s)
+        acct_override(acct_override) {
+    }
 
-    RGWToken token{from_base64(auth_id)};
```
ok

mattbenjamin commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -3936,52 +3939,48 @@ int RGW_Auth_S3::authorize_v2(RGWRados *store, struct req_state *s)
-      } /* success */
-    } /* token */
-  } /* ldap */
+  if (!ldap.is_applicable()) {
```
this looks great

mattbenjamin commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -4342,3 +4341,98 @@ RGWOp* RGWHandler_REST_Service_S3Website::get_obj_op(bool get_data)
+  };
+}
+
+bool RGWLDAPAuthEngine::is_applicable() const noexcept
```
so much better than what we had

mattbenjamin commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -4342,3 +4341,98 @@ RGWOp* RGWHandler_REST_Service_S3Website::get_obj_op(bool get_data)
+}
+
+std::string RGWLDAPTokenExtractor::get_token() const
+{
```
also great

mattbenjamin commented on src/rgw/rgw_rest_s3.h:
```diff
@@ -627,4 +630,48 @@ static inline int valid_s3_bucket_name(const string& name, bool relaxed=false)
+    return "RGWLDAPAuthEngine";
+  }
+  bool is_applicable() const noexcept override;
+  RGWAuthApplier::aplptr_t authenticate() const override;
```
just pointlessly, "aplptr_t" is not a great type/name--I know it's part of the interface...

mattbenjamin commented on src/rgw/rgw_token.h:
```diff
@@ -71,7 +71,7 @@ namespace rgw {
     virtual uint32_t version() const { return 1; };
 
-    bool valid() {
+    bool valid() const{
```
Radek was here? :)

mattbenjamin commented:
The only significant point from my review is, I think, maybe preferring a non-string (perhaps an enumeration type) for account_type?  Past that, is this ready to test?

pritha-srivastava commented:
@mattbenjamin : Reworked the 'type' related code based on your comment. Also, replaced other pieces of 'old' ldap code with the new one. And now have completely removed the old ldap code. This is ready to be tested.

yehudasa commented on src/rgw/rgw_common.h:
```diff
@@ -527,6 +527,14 @@ void encode_json(const char *name, const RGWUserCaps& val, Formatter *f);
+enum RGWUserType
+{
+  TYPE_NONE=0,
+  TYPE_KEYSTONE,
```
@pritha-srivastava I'd put rgw first, then keystone and then ldap. Also, would make the value explicit (even though probably redundant)

yehudasa commented on src/rgw/rgw_common.h:
```diff
@@ -527,6 +527,14 @@ void encode_json(const char *name, const RGWUserCaps& val, Formatter *f);
 
 void decode_json_obj(obj_version& v, JSONObj *obj);
 
+enum RGWUserType
```
a question: is it user type, or user source type?

yehudasa commented on src/rgw/rgw_common.h:
```diff
@@ -605,6 +615,7 @@ struct RGWUserInfo
      ::encode(user_quota, bl);
      ::encode(user_id.tenant, bl);
      ::encode(admin, bl);
+     ::encode(type, bl);
```
need to bump up the encode version in ENCODE_START()

yehudasa commented on src/rgw/rgw_common.h:
```diff
@@ -677,6 +688,7 @@ struct RGWUserInfo
     }
     if (struct_v >= 18) {
       ::decode(admin, bl);
```
`if (struct_v >= 19) {`

yehudasa commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -1705,11 +1705,13 @@ int RGWPostObj_ObjStore_S3::get_policy()
     op_ret = rgw_get_user_info_by_access_key(store, s3_access_key, user_info);
     if (op_ret < 0) {
+      S3AuthFactory aplfact(store, s->account_name);
+      RGWLDAPTokenExtractor token_extr(s);
```
I don't have the larger context of the code, but should we define an RGWLDAPTokenExtractor object if ldap was not configured to be used? maybe different scoping is possible?

pritha-srivastava replied:
RGWLDAPAuthEngine::applicable has all the logic to determine whether ldap is correctly configured to be used or not including checking the configurable. Moreover the extractor will be common to ldap, keystone and local auth, so may end up eventually getting renamed and one object of the extractor will be defined for all three auth engines.(This depends on the review of local auth rework code.)

yehudasa commented on src/rgw/rgw_common.h:
```diff
@@ -547,14 +555,16 @@ struct RGWUserInfo
   RGWQuotaInfo bucket_quota;
   map<int, string> temp_url_keys;
   RGWQuotaInfo user_quota;
+  uint32_t type;
```
@pritha-srivastava also need to update the RGWUserInfo json encoding and decoding

yehudasa replied:
@pritha-srivastava should declare it as RGWUserSourceType?

pritha-srivastava replied:
I have left this as uint32_t since:
1. decode doesn't recognize enums
2. I followed the code for rgw_meta_sync_info and there also the type is uint.

pritha-srivastava commented:
@yehudasa : I have made changes related to the user source type. Can you please re-review.

yehudasa commented on src/rgw/rgw_json_enc.cc:
```diff
@@ -427,6 +427,7 @@ void RGWUserInfo::dump(Formatter *f) const
   encode_json("bucket_quota", bucket_quota, f);
   encode_json("user_quota", user_quota, f);
   encode_json("temp_url_keys", temp_url_keys, f);
+  encode_json("type", type, f);
```
@pritha-srivastava maybe json encode this enum as a string, not as an int (e.g., look at rgw_meta_sync_info)

mattbenjamin commented:
This will need a rebase due to refactoring for rebind().  I can do it, if you like?

mattbenjamin commented:
I've verified the LDAP auth behavior, including rebind.  The one potential problem is related to the new stored user type:  external-auth users already present in a cluster (e.g., RHCS 2.0) will be unable to authenticate.  I worked around this by removing said users.  We need a ruling on whether we need any accommodation (e.g., set user type?)
